File: backend/src/main/java/com/leavemarker/service/PaymentService.java
package com.leavemarker.service;

import com.leavemarker.config.RazorpayConfig;
import com.leavemarker.dto.payment.*;
import com.leavemarker.dto.subscription.SubscriptionRequest;
import com.leavemarker.entity.Company;
import com.leavemarker.entity.Payment;
import com.leavemarker.entity.Plan;
import com.leavemarker.entity.Subscription;
import com.leavemarker.enums.BillingCycle;
import com.leavemarker.enums.PaymentStatus;
import com.leavemarker.enums.PaymentType;
import com.leavemarker.enums.PlanTier;
import com.leavemarker.enums.SubscriptionStatus;
import com.leavemarker.exception.BadRequestException;
import com.leavemarker.exception.ResourceNotFoundException;
import com.leavemarker.repository.CompanyRepository;
import com.leavemarker.repository.EmployeeRepository;
import com.leavemarker.repository.PaymentRepository;
import com.leavemarker.repository.PlanRepository;
import com.leavemarker.repository.SubscriptionRepository;
import com.razorpay.Order;
import com.razorpay.RazorpayClient;
import com.razorpay.RazorpayException;
import com.razorpay.Utils;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.json.JSONObject;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class PaymentService {

    private final PaymentRepository paymentRepository;
    private final SubscriptionRepository subscriptionRepository;
    private final CompanyRepository companyRepository;
    private final PlanRepository planRepository;
    private final EmployeeRepository employeeRepository;
    private final RazorpayConfig razorpayConfig;
    private final RazorpayClient razorpayClient;
    private final SubscriptionService subscriptionService;

    @Transactional(readOnly = true)
    public List<PaymentResponse> getCompanyPayments(Long companyId) {
        Company company = companyRepository.findById(companyId)
                .orElseThrow(() -> new ResourceNotFoundException("Company not found with id: " + companyId));

        return paymentRepository.findByCompany(company).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public PaymentResponse getPaymentById(Long paymentId) {
        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new ResourceNotFoundException("Payment not found with id: " + paymentId));
        return convertToResponse(payment);
    }

    @Transactional
    public PaymentInitiateResponse initiatePayment(Long companyId, PaymentInitiateRequest request, HttpServletRequest httpRequest) {
        Company company = companyRepository.findById(companyId)
                .orElseThrow(() -> new ResourceNotFoundException("Company not found with id: " + companyId));

        Plan plan = planRepository.findById(request.getPlanId())
                .orElseThrow(() -> new ResourceNotFoundException("Plan not found with id: " + request.getPlanId()));

        if (!plan.getActive()) {
            throw new BadRequestException("Selected plan is not active");
        }

        if (plan.getTier() == PlanTier.FREE) {
            throw new BadRequestException("Free plan does not require payment");
        }

        // Check if company already has an active subscription to the same plan
        subscriptionRepository.findByCompanyAndStatus(company, SubscriptionStatus.ACTIVE)
                .ifPresent(existingSubscription -> {
                    if (existingSubscription.getPlan().getId().equals(plan.getId())) {
                        throw new BadRequestException("Company already has an active subscription to this plan");
                    }
                });

        // Generate unique IDs
        String transactionId = "TXN-" + UUID.randomUUID().toString();
        String idempotencyKey = "IDEM-" + UUID.randomUUID().toString();

        LocalDateTime now = LocalDateTime.now();
        LocalDateTime periodEnd = request.getBillingCycle() == BillingCycle.YEARLY
                ? now.plusYears(1)
                : now.plusMonths(1);

        // Calculate amount based on per-employee pricing
        long employeeCount = employeeRepository.countByCompanyIdAndDeletedFalse(companyId);
        if (employeeCount == 0) {
            employeeCount = 1; // Minimum 1 employee for pricing
        }

        BigDecimal pricePerEmployee = request.getBillingCycle() == BillingCycle.YEARLY
                ? plan.getYearlyPrice()
                : plan.getMonthlyPrice();
        BigDecimal amount = pricePerEmployee.multiply(BigDecimal.valueOf(employeeCount));

        // Create payment record WITHOUT subscription (subscription created after payment success)
        Payment payment = Payment.builder()
                .plan(plan)
                .company(company)
                .transactionId(transactionId)
                .paymentType(PaymentType.NEW_SUBSCRIPTION)
                .billingCycle(request.getBillingCycle())
                .amount(amount)
                .taxAmount(BigDecimal.ZERO)
                .discountAmount(BigDecimal.ZERO)
                .totalAmount(amount)
                .currency("INR")
                .status(PaymentStatus.PENDING)
                .initiatedAt(now)
                .periodStart(now)
                .periodEnd(periodEnd)
                .idempotencyKey(idempotencyKey)
                .ipAddress(getClientIp(httpRequest))
                .userAgent(httpRequest != null ? httpRequest.getHeader("User-Agent") : null)
                .build();

        Payment savedPayment = paymentRepository.save(payment);

        try {
            // Create Razorpay Order
            String razorpayOrderId = createRazorpayOrder(savedPayment);
            savedPayment.setRazorpayOrderId(razorpayOrderId);
            paymentRepository.save(savedPayment);

            log.info("Payment initiated for company: {}, plan: {}, transaction: {}, razorpayOrderId: {}",
                    companyId, plan.getName(), transactionId, razorpayOrderId);

            return PaymentInitiateResponse.builder()
                    .razorpayOrderId(razorpayOrderId)
                    .razorpayKeyId(razorpayConfig.getKeyId())
                    .amount(savedPayment.getTotalAmount().multiply(BigDecimal.valueOf(100)).longValue())
                    .currency(savedPayment.getCurrency())
                    .transactionId(transactionId)
                    .companyName(company.getName())
                    .companyEmail(company.getEmail())
                    .employeeCount(employeeCount)
                    .pricePerEmployee(pricePerEmployee.multiply(BigDecimal.valueOf(100)).longValue())
                    .build();

        } catch (Exception e) {
            savedPayment.setStatus(PaymentStatus.FAILED);
            savedPayment.setFailedAt(now);
            savedPayment.setFailureReason("Failed to create payment: " + e.getMessage());
            paymentRepository.save(savedPayment);
            log.error("Failed to initiate payment for transaction: {}", transactionId, e);
            throw new BadRequestException("Failed to initiate payment: " + e.getMessage());
        }
    }

    @Transactional
    public PaymentResponse verifyAndCompletePayment(PaymentVerifyRequest verifyRequest) {
        Payment payment = paymentRepository.findByRazorpayOrderId(verifyRequest.getRazorpayOrderId())
                .orElseThrow(() -> new ResourceNotFoundException("Payment not found with razorpayOrderId: " + verifyRequest.getRazorpayOrderId()));

        LocalDateTime now = LocalDateTime.now();

        try {
            // Verify signature
            JSONObject options = new JSONObject();
            options.put("razorpay_order_id", verifyRequest.getRazorpayOrderId());
            options.put("razorpay_payment_id", verifyRequest.getRazorpayPaymentId());
            options.put("razorpay_signature", verifyRequest.getRazorpaySignature());

            boolean isValidSignature = Utils.verifyPaymentSignature(options, razorpayConfig.getKeySecret());

            if (!isValidSignature) {
                payment.setStatus(PaymentStatus.FAILED);
                payment.setFailedAt(now);
                payment.setFailureReason("Invalid payment signature");
                paymentRepository.save(payment);
                throw new BadRequestException("Invalid payment signature");
            }

            // Update payment with Razorpay details
            payment.setRazorpayPaymentId(verifyRequest.getRazorpayPaymentId());
            payment.setRazorpaySignature(verifyRequest.getRazorpaySignature());
            payment.setStatus(PaymentStatus.SUCCESS);
            payment.setPaidAt(now);

            // Create subscription ONLY after successful payment verification
            SubscriptionRequest subscriptionRequest = new SubscriptionRequest();
            subscriptionRequest.setPlanId(payment.getPlan().getId());
            subscriptionRequest.setBillingCycle(payment.getBillingCycle());
            subscriptionRequest.setAutoRenew(true);

            var subscriptionResponse = subscriptionService.createSubscription(
                    payment.getCompany().getId(), subscriptionRequest);

            // Link the subscription to the payment
            Subscription subscription = subscriptionRepository.findById(subscriptionResponse.getId())
                    .orElseThrow(() -> new ResourceNotFoundException("Subscription not found"));
            subscription.setIsPaid(true);
            subscription.setCurrentPeriodEnd(payment.getPeriodEnd());
            subscription.setEndDate(payment.getPeriodEnd());
            subscriptionRepository.save(subscription);

            payment.setSubscription(subscription);
            paymentRepository.save(payment);

            log.info("Payment verified and subscription created for transaction: {}, razorpayPaymentId: {}",
                    payment.getTransactionId(), verifyRequest.getRazorpayPaymentId());

            return convertToResponse(payment);

        } catch (RazorpayException e) {
            payment.setStatus(PaymentStatus.FAILED);
            payment.setFailedAt(now);
            payment.setFailureReason("Signature verification failed: " + e.getMessage());
            paymentRepository.save(payment);
            log.error("Payment verification failed for transaction: {}", payment.getTransactionId(), e);
            throw new BadRequestException("Payment verification failed: " + e.getMessage());
        }
    }

    @Transactional
    public void handleWebhook(String rawPayload, String razorpaySignature) {
        try {
            // Verify webhook signature
            boolean isValid = Utils.verifyWebhookSignature(rawPayload, razorpaySignature, razorpayConfig.getWebhookSecret());

            if (!isValid) {
                log.warn("Invalid webhook signature received");
                throw new BadRequestException("Invalid webhook signature");
            }

            JSONObject webhookPayload = new JSONObject(rawPayload);
            String event = webhookPayload.getString("event");
            JSONObject payloadData = webhookPayload.getJSONObject("payload");
            JSONObject paymentEntity = payloadData.getJSONObject("payment").getJSONObject("entity");

            String razorpayOrderId = paymentEntity.getString("order_id");
            String razorpayPaymentId = paymentEntity.getString("id");

            Payment payment = paymentRepository.findByRazorpayOrderId(razorpayOrderId)
                    .orElse(null);

            if (payment == null) {
                log.warn("Payment not found for razorpayOrderId: {}", razorpayOrderId);
                return;
            }

            LocalDateTime now = LocalDateTime.now();
            payment.setWebhookStatus(event);
            payment.setWebhookReceivedAt(now);
            payment.setWebhookPayload(rawPayload);
            payment.setRazorpayPaymentId(razorpayPaymentId);

            switch (event) {
                case "payment.captured" -> {
                    if (payment.getStatus() != PaymentStatus.SUCCESS) {
                        payment.setStatus(PaymentStatus.SUCCESS);
                        payment.setPaidAt(now);

                        String method = paymentEntity.optString("method", null);
                        payment.setPaymentMethod(method);

                        // Create subscription if not exists (webhook came before verify call)
                        if (payment.getSubscription() == null) {
                            SubscriptionRequest subscriptionRequest = new SubscriptionRequest();
                            subscriptionRequest.setPlanId(payment.getPlan().getId());
                            subscriptionRequest.setBillingCycle(payment.getBillingCycle());
                            subscriptionRequest.setAutoRenew(true);

                            var subscriptionResponse = subscriptionService.createSubscription(
                                    payment.getCompany().getId(), subscriptionRequest);

                            Subscription subscription = subscriptionRepository.findById(subscriptionResponse.getId())
                                    .orElseThrow(() -> new ResourceNotFoundException("Subscription not found"));
                            subscription.setIsPaid(true);
                            subscription.setCurrentPeriodEnd(payment.getPeriodEnd());
                            subscription.setEndDate(payment.getPeriodEnd());
                            subscriptionRepository.save(subscription);

                            payment.setSubscription(subscription);
                        } else {
                            // Update existing subscription
                            Subscription subscription = payment.getSubscription();
                            subscription.setIsPaid(true);
                            subscription.setCurrentPeriodEnd(payment.getPeriodEnd());
                            subscription.setEndDate(payment.getPeriodEnd());
                            subscription.setStatus(SubscriptionStatus.ACTIVE);
                            subscriptionRepository.save(subscription);
                        }

                        log.info("Payment captured via webhook for transaction: {}", payment.getTransactionId());
                    }
                }
                case "payment.failed" -> {
                    payment.setStatus(PaymentStatus.FAILED);
                    payment.setFailedAt(now);

                    JSONObject errorObj = paymentEntity.optJSONObject("error_description");
                    String errorReason = errorObj != null ? errorObj.optString("description", "Payment failed") : "Payment failed";
                    payment.setFailureReason(errorReason);

                    log.warn("Payment failed via webhook for transaction: {}", payment.getTransactionId());
                }
                case "refund.created" -> {
                    payment.setStatus(PaymentStatus.REFUNDED);
                    payment.setRefundedAt(now);
                    log.info("Payment refunded via webhook for transaction: {}", payment.getTransactionId());
                }
                default -> log.info("Received webhook event: {} for transaction: {}", event, payment.getTransactionId());
            }

            paymentRepository.save(payment);

        } catch (RazorpayException e) {
            log.error("Webhook signature verification failed", e);
            throw new BadRequestException("Invalid webhook signature");
        } catch (Exception e) {
            log.error("Error processing webhook", e);
            throw new BadRequestException("Error processing webhook: " + e.getMessage());
        }
    }

    @Transactional
    public PaymentInitiateResponse retryPayment(Long paymentId) {
        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new ResourceNotFoundException("Payment not found"));

        if (payment.getStatus() != PaymentStatus.FAILED) {
            throw new BadRequestException("Can only retry failed payments");
        }

        if (payment.getRetryCount() >= 3) {
            throw new BadRequestException("Maximum retry attempts exceeded");
        }

        payment.setRetryCount(payment.getRetryCount() + 1);
        payment.setLastRetryAt(LocalDateTime.now());
        payment.setStatus(PaymentStatus.PENDING);

        try {
            String razorpayOrderId = createRazorpayOrder(payment);
            payment.setRazorpayOrderId(razorpayOrderId);
            paymentRepository.save(payment);

            log.info("Payment retry initiated for transaction: {}, attempt: {}",
                    payment.getTransactionId(), payment.getRetryCount());

            return PaymentInitiateResponse.builder()
                    .razorpayOrderId(razorpayOrderId)
                    .razorpayKeyId(razorpayConfig.getKeyId())
                    .amount(payment.getTotalAmount().multiply(BigDecimal.valueOf(100)).longValue())
                    .currency(payment.getCurrency())
                    .transactionId(payment.getTransactionId())
                    .companyName(payment.getCompany().getName())
                    .companyEmail(payment.getCompany().getEmail())
                    .build();

        } catch (Exception e) {
            payment.setStatus(PaymentStatus.FAILED);
            payment.setFailureReason("Retry failed: " + e.getMessage());
            paymentRepository.save(payment);
            throw new BadRequestException("Payment retry failed: " + e.getMessage());
        }
    }

    private String createRazorpayOrder(Payment payment) throws RazorpayException {
        JSONObject orderRequest = new JSONObject();
        orderRequest.put("amount", payment.getTotalAmount().multiply(BigDecimal.valueOf(100)).longValue()); // Convert to paise
        orderRequest.put("currency", payment.getCurrency());
        orderRequest.put("receipt", payment.getTransactionId());

        // Add notes for tracking
        JSONObject notes = new JSONObject();
        notes.put("company_id", payment.getCompany().getId().toString());
        notes.put("plan_id", payment.getPlan().getId().toString());
        notes.put("billing_cycle", payment.getBillingCycle().toString());
        notes.put("transaction_id", payment.getTransactionId());
        orderRequest.put("notes", notes);

        Order order = razorpayClient.orders.create(orderRequest);
        return order.get("id");
    }

    private String getClientIp(HttpServletRequest request) {
        if (request == null) return null;

        String xForwardedFor = request.getHeader("X-Forwarded-For");
        if (xForwardedFor != null && !xForwardedFor.isEmpty()) {
            return xForwardedFor.split(",")[0].trim();
        }
        return request.getRemoteAddr();
    }

    private PaymentResponse convertToResponse(Payment payment) {
        return PaymentResponse.builder()
                .id(payment.getId())
                .subscriptionId(payment.getSubscription() != null ? payment.getSubscription().getId() : null)
                .transactionId(payment.getTransactionId())
                .razorpayOrderId(payment.getRazorpayOrderId())
                .razorpayPaymentId(payment.getRazorpayPaymentId())
                .amount(payment.getTotalAmount())
                .currency(payment.getCurrency())
                .status(payment.getStatus())
                .paymentMethod(payment.getPaymentMethod())
                .paidAt(payment.getPaidAt())
                .failureReason(payment.getFailureReason())
                .createdAt(payment.getCreatedAt())
                .build();
    }
}

File: backend-go/internal/service/payment_service.go
package service

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/leavemarker/backend-go/internal/config"
	"github.com/leavemarker/backend-go/internal/dto"
	"github.com/leavemarker/backend-go/internal/enums"
	"github.com/leavemarker/backend-go/internal/models"
	"github.com/leavemarker/backend-go/internal/repository"
	"github.com/leavemarker/backend-go/internal/utils"
	razorpay "github.com/razorpay/razorpay-go"
	"github.com/shopspring/decimal"
)

// PaymentService handles payment business logic
type PaymentService struct {
	paymentRepo      *repository.PaymentRepository
	subscriptionRepo *repository.SubscriptionRepository
	planRepo         *repository.PlanRepository
	razorpayClient   *razorpay.Client
	cfg              *config.Config
}

// NewPaymentService creates a new PaymentService
func NewPaymentService(
	paymentRepo *repository.PaymentRepository,
	subscriptionRepo *repository.SubscriptionRepository,
	planRepo *repository.PlanRepository,
	cfg *config.Config,
) *PaymentService {
	var client *razorpay.Client
	if cfg.Razorpay.KeyID != "" && cfg.Razorpay.KeySecret != "" {
		client = razorpay.NewClient(cfg.Razorpay.KeyID, cfg.Razorpay.KeySecret)
	}
	return &PaymentService{
		paymentRepo:      paymentRepo,
		subscriptionRepo: subscriptionRepo,
		planRepo:         planRepo,
		razorpayClient:   client,
		cfg:              cfg,
	}
}

// GetPayments gets all payments for a company
func (s *PaymentService) GetPayments(companyID uint) ([]dto.PaymentResponse, error) {
	payments, err := s.paymentRepo.FindByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to get payments")
	}

	responses := make([]dto.PaymentResponse, len(payments))
	for i, payment := range payments {
		responses[i] = *s.toPaymentResponse(&payment)
	}

	return responses, nil
}

// GetPayment gets a payment by ID
func (s *PaymentService) GetPayment(id, companyID uint) (*dto.PaymentResponse, error) {
	payment, err := s.paymentRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Payment not found")
	}

	if payment.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	return s.toPaymentResponse(payment), nil
}

// InitiatePayment initiates a new payment
func (s *PaymentService) InitiatePayment(companyID uint, req dto.PaymentInitiateRequest, ipAddress, userAgent string) (*dto.PaymentInitiateResponse, error) {
	if s.razorpayClient == nil {
		return nil, utils.NewInternalServerError("Payment gateway not configured")
	}

	plan, err := s.planRepo.FindByID(req.PlanID)
	if err != nil {
		return nil, utils.NewNotFoundError("Plan not found")
	}

	// Calculate amount
	var amount decimal.Decimal
	if req.BillingCycle == enums.BillingCycleMonthly {
		amount = plan.MonthlyPrice
	} else {
		amount = plan.YearlyPrice
	}

	// Convert to paise (Razorpay uses smallest currency unit)
	amountInPaise := amount.Mul(decimal.NewFromInt(100)).IntPart()

	// Create Razorpay order
	orderData := map[string]interface{}{
		"amount":   amountInPaise,
		"currency": "INR",
		"receipt":  fmt.Sprintf("rcpt_%s", uuid.New().String()[:8]),
	}

	order, err := s.razorpayClient.Order.Create(orderData, nil)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to create payment order")
	}

	orderID := order["id"].(string)

	// Calculate period
	now := time.Now()
	var periodEnd time.Time
	if req.BillingCycle == enums.BillingCycleMonthly {
		periodEnd = now.AddDate(0, 1, 0)
	} else {
		periodEnd = now.AddDate(1, 0, 0)
	}

	// Create payment record
	transactionID := uuid.New().String()
	payment := &models.Payment{
		PlanID:          req.PlanID,
		CompanyID:       companyID,
		TransactionID:   transactionID,
		RazorpayOrderID: &orderID,
		PaymentType:     enums.PaymentTypeSubscription,
		BillingCycle:    req.BillingCycle,
		Amount:          amount,
		TotalAmount:     amount,
		Currency:        "INR",
		Status:          enums.PaymentStatusPending,
		InitiatedAt:     &now,
		PeriodStart:     now,
		PeriodEnd:       periodEnd,
		IPAddress:       &ipAddress,
		UserAgent:       &userAgent,
	}

	if err := s.paymentRepo.Create(payment); err != nil {
		return nil, utils.NewInternalServerError("Failed to create payment record")
	}

	return &dto.PaymentInitiateResponse{
		OrderID:  orderID,
		Amount:   amount,
		Currency: "INR",
		KeyID:    s.cfg.Razorpay.KeyID,
	}, nil
}

// VerifyPayment verifies and completes a payment
func (s *PaymentService) VerifyPayment(companyID uint, req dto.PaymentVerifyRequest) (*dto.PaymentResponse, error) {
	// Verify signature
	message := req.RazorpayOrderID + "|" + req.RazorpayPaymentID
	expectedSignature := s.generateSignature(message, s.cfg.Razorpay.KeySecret)

	if expectedSignature != req.RazorpaySignature {
		return nil, utils.NewBadRequestError("Invalid payment signature")
	}

	// Find payment by order ID
	payment, err := s.paymentRepo.FindByRazorpayOrderID(req.RazorpayOrderID)
	if err != nil {
		return nil, utils.NewNotFoundError("Payment not found")
	}

	if payment.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	// Update payment
	now := time.Now()
	payment.RazorpayPaymentID = &req.RazorpayPaymentID
	payment.RazorpaySignature = &req.RazorpaySignature
	payment.Status = enums.PaymentStatusSuccess
	payment.PaidAt = &now

	if err := s.paymentRepo.Update(payment); err != nil {
		return nil, utils.NewInternalServerError("Failed to update payment")
	}

	// Activate subscription
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err == nil && subscription != nil {
		subscription.IsPaid = true
		subscription.PlanID = payment.PlanID
		subscription.Amount = payment.Amount
		subscription.StartDate = payment.PeriodStart
		subscription.EndDate = payment.PeriodEnd
		subscription.CurrentPeriodStart = payment.PeriodStart
		subscription.CurrentPeriodEnd = payment.PeriodEnd
		_ = s.subscriptionRepo.Update(subscription)
	}

	return s.toPaymentResponse(payment), nil
}

// HandleWebhook handles Razorpay webhook
func (s *PaymentService) HandleWebhook(payload []byte, signature string) error {
	// Verify webhook signature
	expectedSignature := s.generateSignature(string(payload), s.cfg.Razorpay.WebhookSecret)
	if expectedSignature != signature {
		return utils.NewBadRequestError("Invalid webhook signature")
	}

	// Process webhook (simplified - in production, parse JSON and handle different events)
	return nil
}

// RetryPayment retries a failed payment
func (s *PaymentService) RetryPayment(id, companyID uint) (*dto.PaymentInitiateResponse, error) {
	payment, err := s.paymentRepo.FindByID(id)
	if err != nil {
		return nil, utils.NewNotFoundError("Payment not found")
	}

	if payment.CompanyID != companyID {
		return nil, utils.NewForbiddenError("Access denied")
	}

	if payment.Status != enums.PaymentStatusFailed {
		return nil, utils.NewBadRequestError("Only failed payments can be retried")
	}

	// Create new Razorpay order
	amountInPaise := payment.Amount.Mul(decimal.NewFromInt(100)).IntPart()
	orderData := map[string]interface{}{
		"amount":   amountInPaise,
		"currency": "INR",
		"receipt":  fmt.Sprintf("retry_%s", uuid.New().String()[:8]),
	}

	order, err := s.razorpayClient.Order.Create(orderData, nil)
	if err != nil {
		return nil, utils.NewInternalServerError("Failed to create payment order")
	}

	orderID := order["id"].(string)

	// Update payment record
	now := time.Now()
	payment.RazorpayOrderID = &orderID
	payment.Status = enums.PaymentStatusPending
	payment.RetryCount++
	payment.LastRetryAt = &now

	if err := s.paymentRepo.Update(payment); err != nil {
		return nil, utils.NewInternalServerError("Failed to update payment record")
	}

	return &dto.PaymentInitiateResponse{
		OrderID:  orderID,
		Amount:   payment.Amount,
		Currency: "INR",
		KeyID:    s.cfg.Razorpay.KeyID,
	}, nil
}

func (s *PaymentService) generateSignature(message, secret string) string {
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(message))
	return hex.EncodeToString(h.Sum(nil))
}

func (s *PaymentService) toPaymentResponse(payment *models.Payment) *dto.PaymentResponse {
	response := &dto.PaymentResponse{
		ID:                payment.ID,
		TransactionID:     payment.TransactionID,
		PlanID:            payment.PlanID,
		SubscriptionID:    payment.SubscriptionID,
		PaymentType:       payment.PaymentType,
		BillingCycle:      payment.BillingCycle,
		Amount:            payment.Amount,
		TaxAmount:         payment.TaxAmount,
		DiscountAmount:    payment.DiscountAmount,
		TotalAmount:       payment.TotalAmount,
		Currency:          payment.Currency,
		Status:            payment.Status,
		PaymentMethod:     payment.PaymentMethod,
		RazorpayOrderID:   payment.RazorpayOrderID,
		RazorpayPaymentID: payment.RazorpayPaymentID,
		PaidAt:            payment.PaidAt,
		PeriodStart:       payment.PeriodStart,
		PeriodEnd:         payment.PeriodEnd,
		CreatedAt:         payment.CreatedAt,
		UpdatedAt:         payment.UpdatedAt,
	}

	if payment.Plan.ID != 0 {
		response.PlanName = payment.Plan.Name
	}

	return response
}
